src/clp_ffi_py/ir/native/PyDeserializer.cpp

```diff
@@ -0,0 +1,281 @@
+auto PyDeserializer::init(
+        PyObject* input_stream,
+        Py_ssize_t buffer_capacity,
+        bool allow_incomplete_stream
+) -> bool {
+    m_allow_incomplete_stream = allow_incomplete_stream;
+    m_deserializer_buffer_reader = DeserializerBufferReader::create(input_stream, buffer_capacity);
+    if (nullptr == m_deserializer_buffer_reader) {
+        return false;
+    }
+
+    try {
+        PyDeserializer::IrUnitHandler::LogEventHandle log_event_handle
+                = [this](clp::ffi::KeyValuePairLogEvent&& kv_log_event) -> IRErrorCode {
+            return this->handle_log_event(std::move(kv_log_event));
+        };
+
+        PyDeserializer::IrUnitHandler::UtcOffsetChangeHandle trivial_utc_offset_handle
+                = []([[maybe_unused]] clp::UtcOffset, [[maybe_unused]] clp::UtcOffset
+                  ) -> IRErrorCode { return IRErrorCode::IRErrorCode_Success; };
+
+        PyDeserializer::IrUnitHandler::SchemaTreeNodeInsertionHandle
+                trivial_schema_tree_node_insertion_handle
+                = []([[maybe_unused]] clp::ffi::SchemaTree::NodeLocator) -> IRErrorCode {
+            return IRErrorCode::IRErrorCode_Success;
+        };
+
+        PyDeserializer::IrUnitHandler::EndOfStreamHandle end_of_stream_handle
+                = [this]() -> IRErrorCode { return this->handle_end_of_stream(); };
+
+        auto deserializer_result{Deserializer::create(
+                *m_deserializer_buffer_reader,
+                {std::move(log_event_handle),
+                 std::move(trivial_utc_offset_handle),
+                 std::move(trivial_schema_tree_node_insertion_handle),
+                 std::move(end_of_stream_handle)}
+        )};
+        if (deserializer_result.has_error()) {
+            PyErr_Format(
+                    PyExc_RuntimeError,
+                    cDeserializerCreateErrorFormatStr.data(),
+                    deserializer_result.error().message().c_str()
+            );
+            return false;
+        }
+        m_deserializer = new clp::ffi::ir_stream::Deserializer<PyDeserializer::IrUnitHandler>{
+                std::move(deserializer_result.value())
+        };
+    } catch (clp::TraceableException& exception) {
+        handle_traceable_exception(exception);
+        return false;
+    }
+
+    return true;
+}
+
+auto PyDeserializer::deserialize_log_event() -> PyObject* {
+    try {
+        while (false == is_stream_complete()) {
+            auto const ir_unit_type_result{
+                    m_deserializer->deserialize_next_ir_unit(*m_deserializer_buffer_reader)
+            };
+            if (ir_unit_type_result.has_error()) {
+                if (false == handle_incomplete_ir_error(ir_unit_type_result.error())) {
+                    return nullptr;
+                }
+                break;
+            }
+            auto const ir_unit_type{ir_unit_type_result.value()};
+            if (IrUnitType::LogEvent == ir_unit_type && has_unreleased_deserialized_log_event()) {
```

Comment:
Just to help me understand the code,

what would be a case where
IrUnitType::LogEvent == ir_unit_type is True, but has_unreleased_deserialized_log_event is False

Reply:
thinking again. I think we don't need to check the deserialized IR unit type at all. We just need to check whether it returns an error or not.
But there could be cases where the return IR unit type is `LogEvent` without unreleased log event buffered. If we implement native query, only matched log event will be stored.

Reply:
>But there could be cases where the return IR unit type is LogEvent without unreleased log event buffered. If we implement native query, only matched log event will be stored.

Do we expect the native query to be implemented soon? If it's not in the near future, I think we don't have to prepare for this case in the code, as it would cause some confusions.

It makes more sense to me to add the check for has_unreleased_deserialized_log_event() only when it is needed, or at least we should add a comment.

Reply:
I have refactored this piece of code.
In general, I think:
- We should check the returned IR unit type as it's the expected usage of the core deserializer API
- We should also ensure there is an unreleased deserialized log event before we try to "release" this pointer

That said, I think both checks are necessary, but we can write it so that it doesn't look confusing.

Reply:
Disuccsed offline. The latest commit should reflect our conclusion:
- We should keep both check as discussed above
- We should check `has_unreleased_deserialized_log_event` to meet the requirement that the ownership must be validated by the caller before calling `release_deserialized_log_event`
- If the validation failed, we should throw a Python exception instead of `continue`. `continue` logic should be implemented in the PR which native queries are implemented